### DISCUSSION TITLE
How to get CUDA to work?

### DISCUSSION THREAD

**Discussion by thhart** [deepjavalibrary/djl#2505] [Q&A] [answered]
Hi, I have problems to get cuda support running, CUDA libraries are available as 11.7 and 12.0, however neither mxnet nor pytorch  can find the correct cuda version. Always I get a error message like this: 

PyTorch:
"cuda flavor for linux-x86_64 found: cu120."
MXnet:
"No matching cuda flavor for linux found: cu120mkl/sm_75."

Other libraries like onnx native (outside ai.djl) have no problems finding and using CUDA.

pom.xml contains this:

        <dependency>
            <groupId>ai.djl</groupId>
            <artifactId>api</artifactId>
            <version>0.21.0</version>
        </dependency>
        
        <dependency>
            <groupId>ai.djl.mxnet</groupId>
            <artifactId>mxnet-engine</artifactId>
            <version>0.21.0</version>
            <scope>runtime</scope>
        </dependency>
        
        <dependency>
            <groupId>ai.djl.pytorch</groupId>
            <artifactId>pytorch-engine</artifactId>
            <version>0.21.0</version>
        </dependency>
        
        <dependency>
            <groupId>ai.djl.mxnet</groupId>
            <artifactId>mxnet-native-cu110mkl</artifactId>
            <version>1.9.0</version>
        </dependency>
        <dependency>
            <groupId>ai.djl.pytorch</groupId>
            <artifactId>pytorch-jni</artifactId>
            <version>1.13.1-0.21.0</version>
            <scope>runtime</scope>
        </dependency>
        <dependency>
            <groupId>ai.djl.pytorch</groupId>
            <classifier>linux-x86_64</classifier>
            <artifactId>pytorch-native-cu117</artifactId>
            <version>1.13.1</version>
            <scope>runtime</scope>
        </dependency>

**Comment by frankfliu** [comment, marked as answer]
MXNext only works for cu102 and cu112, see: https://docs.djl.ai/master/engines/mxnet/mxnet-engine/index.html#linux-gpu

PyTorch should work on cu117, what you need to do is set current cuda version to 11.7. DJL try to load default `libcudart.so` file and detect the current cuda version. Make sure cu117 version of libcudart.so is before cu120.

If you use DJL 0.22.0-SNAPSHOT version, you should be able to use cuda 12.0. See: https://github.com/deepjavalibrary/djl/issues/2485

**Comment by thhart** [comment]
Thanks for the pointer, 11.2 is working and finding the GPU with mxnet.
